src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -602,8 +603,62 @@ namespace QgsWms
         }
       }
 
+      QList<QgsMapLayer *> renderLayers, auxHighlightLayers;
+
       if ( !map->keepLayerSet() )
       {
+
+        static const QRegularExpression RE_HIGHLIGHT_LAYER = QRegularExpression( "^highlight_\\d+$" );
+
+        QString presetName = map->followVisibilityPresetName();
+
+        QString newPresetName;
+        if ( !presetName.isEmpty() )
+        {
+          // get layers from theme
+          if ( c->project()->mapThemeCollection()->hasMapTheme( presetName ) )
+            renderLayers = c->project()->mapThemeCollection()->mapThemeVisibleLayers( presetName );
+          else
+            renderLayers = c->project()->mapThemeCollection()->masterVisibleLayers();
+
+          // get highlight layers, if any
+          bool has_highlight_layer = false;
+          for ( auto layer : mapSettings.layers() )
+          {
+
+            if ( RE_HIGHLIGHT_LAYER.match( layer->name() ).hasMatch() )
+            {
+              has_highlight_layer = true;
+              auxHighlightLayers << layer;
+              c->project()->layerTreeRoot()->insertLayer( 0, layer );
+            }
+
+          }
+
+          if ( has_highlight_layer )
+          {
+
+            // create a new theme with all layers
+            QgsMapThemeCollection::MapThemeRecord rec;
+            // layers from the theme
+            for ( auto layer : renderLayers )
```

Comment:
```suggestion
            for ( auto layer : qgis::as_const( renderLayers ) )
```

When iterating containers, make sure they are const